### PR TITLE
Rdma proxy client fixes

### LINKED ISSUE #317: RDMA Proxy client: infinite loop reading the proxy response
Observed during tests:
```
void RDMAProxy_client::_really_read(char *buf, const int len) {
  // ... snip ...
  while (todo > 0 && _request_time_left > 0) {
    double t0 = _stamp_ms();

    // ... snip ... 
    int rc = rpoll(&pollfd, nfds, _request_time_left);
    
    // ... snip ...
    read = rrecv(_socket, &buf[off], todo, flags);
    if (read < 0) {
      throw proxy_exception(read, _build_msg("really_read.rrecv"));
    }

     // ... snip ...
    double t1 = _stamp_ms();
    double delta = t1 - t0;
    _request_time_left = _request_time_left - (int)delta;
  }
}
```

The case `read == 0` is not handled correctly:
> The return value  will  be  0 when the peer has performed an orderly shutdown

(from `recv(2)`).

Together with the rounding caused by `_stamp_ms()` and the `double` -> `int` conversion, this can cause `_request_time_left` to stay at its current value forever.

### PR DESCRIPTION
**Pull request by domsj**
Should fix #317 & #318 

@redlicha can you review?

(will test on 154.231)

### CODE REVIEW

File: cpp/src/lib/rdma_proxy_client.cc [hunk 1]
```diff
@@ -41,64 +42,90 @@ std::string _build_msg(const std::string &prefix) {
 }
 
 void RDMAProxy_client::_really_write(const char *buf, const int len) {
+  if (len <= 0) {
+    return;
```
Comment (redlicha) [inline cpp/src/lib/rdma_proxy_client.cc]: Meta question as this is not something newly introduced: is `len < 0` something that the code should tolerate, or rather something to complain about via `assert` / exception, or to be ruled out altogether with an unsigned typed?

File: cpp/src/lib/rdma_proxy_client.cc [hunk 2]
```diff
@@ -41,64 +42,90 @@ std::string _build_msg(const std::string &prefix) {
...
+
   int flags = 0;
   int sent;
   int todo = len;
   int off = 0;
   int nfds = 1;
 
-  while (todo > 0 && _request_time_left > 0) {
-    double t0 = _stamp_ms();
-    ALBA_LOG(DEBUG, "todo=" << todo
```
Comment (redlicha) [inline cpp/src/lib/rdma_proxy_client.cc]: `_stamp_ms()` is unused now AFAICS and can be removed.

File: cpp/src/lib/rdma_proxy_client.cc [hunk 3]
```diff
@@ -41,64 +42,90 @@ std::string _build_msg(const std::string &prefix) {
...
-    double t1 = _stamp_ms();
-    double delta = t1 - t0;
-    _request_time_left = _request_time_left - (int)delta;
-  }
+  } while (todo > 0 && time_remaining.count() > 0);
+
+  time_remaining = end - std::chrono::steady_clock::now();
+  _request_time_left -= std::chrono::duration_cast<std::chrono::milliseconds>(
+                            time_remaining).count();
 }
```
Comment (redlicha) [inline cpp/src/lib/rdma_proxy_client.cc]: I think replacing `RDMAProxy_client`'s `int request_time_left /* ms */` member with a `chrono::steady_clock::time_point _deadline` member could simplify all of this by avoiding a few conversions between `int`, `time_point` and `duration`, e.g.
```
while (// there is data to be sent)
{
    const auto now = chrono::steady_clock::now();
    if (now >= _deadline)
    {
          // throw timeout exception
    }
    const auto time_left = _deadline - now;
    poll(..., duration_cast<milliseconds>(time_left).count());
    // handle poll errors
    send(...);
    // handle send errors
}
````

File: cpp/src/lib/rdma_proxy_client.cc (hunk 2 above)
Comment (domsj) [inline cpp/src/lib/rdma_proxy_client.cc]: indeed forgot to remove this

File: cpp/src/lib/rdma_proxy_client.cc (hunk 3 above)
Comment (domsj) [inline cpp/src/lib/rdma_proxy_client.cc]: ok

File: cpp/src/lib/rdma_proxy_client.cc (hunk 1 above)
Comment (domsj) [inline cpp/src/lib/rdma_proxy_client.cc]: ok, will throw an exception there

**Comment by toolslive** [comment]
fine for me.

**Pull request merged**